File: api/src/main/java/ru/endlesscode/mimic/api/system/ClassSystem.java
/*
 * This file is part of MimicAPI.
 * Copyright (C) 2018 Osip Fatkullin
 * Copyright (C) 2018 EndlessCode Group and contributors
 *
 * MimicAPI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * MimicAPI is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with MimicAPI.  If not, see <http://www.gnu.org/licenses/>.
 */

package ru.endlesscode.mimic.api.system;

import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.function.Function;

/**
 * System that provides methods to work with players class systems.
 *
 * <p>Before implementing run an eye over all default method implementations
 * and override all methods that works not properly for your case.</p>
 *
 * <p>See {@link PlayerSystem} To read more about implementation.</p>
 *
 * @author Osip Fatkullin
 * @since 0.1
 */
public abstract class ClassSystem implements PlayerSystem {
    /**
     * Checks player has any class.
     *
     * @return {@code true} if player has any class
     */
    public boolean hasClass() {
        return !this.getPrimaryClass().isEmpty();
    }

    /**
     * Checks player has one of required classes.
     *
     * @param requiredClasses List of required classes
     * @return {@code true} if player has one of required class
     */
    public boolean hasOneOfRequiredClasses(@NotNull List<String> requiredClasses) {
        for (String requiredClass : requiredClasses) {
            if (this.hasRequiredClass(requiredClass)) {
                return true;
            }
        }

        return false;
    }

    /**
     * Checks player has all required classes.
     *
     * @param requiredClasses List of required classes
     * @return {@code true} if player has all required class
     */
    public boolean hasAllRequiredClasses(@NotNull List<String> requiredClasses) {
        for (String requiredClass : requiredClasses) {
            if (!this.hasRequiredClass(requiredClass)) {
                return false;
            }
        }

        return true;
    }

    /**
     * Checks player has required class.
     *
     * @param requiredClass Required class name
     * @return {@code true} if player has required class
     */
    public boolean hasRequiredClass(@NotNull String requiredClass) {
        return this.getClasses().contains(requiredClass);
    }

    /**
     * Gets primary class for player.
     *
     * @apiNote
     * Class is called "primary" because some systems can support many system
     * for one player.
     *
     * @implSpec
     * Method shouldn't return {@code null}, but can return empty {@code String}
     *
     * @return Primary class name
     */
    public @NotNull String getPrimaryClass() {
        List<String> playerClasses = this.getClasses();
        return playerClasses.isEmpty() ? "" : playerClasses.get(0);
    }

    /**
     * Gets {@code Lost} of player system.
     *
     * @apiNote
     * This method actual for systems which support many system for one player.
     * If system not support - it just return {@code List} with one element.
     *
     * @implSpec
     * Method shouldn't return {@code null}, but can return empty {@code List}.
     * Also must not contain null objects.
     *
     * @return {@code List} of player system names
     * @throws IllegalStateException If player-related object not exists.
     */
    public abstract @NotNull List<String> getClasses();

    /**
     * Factory of class systems.
     */
    public static class Factory extends SystemFactory<ClassSystem> {
        public Factory(Function<Object, ? extends ClassSystem> constructor) {
            super(constructor);
        }
    }
}

File: api/src/main/kotlin/ru/endlesscode/mimic/api/system/ClassSystem.kt
/*
 * This file is part of MimicAPI.
 * Copyright (C) 2018 Osip Fatkullin
 * Copyright (C) 2018 EndlessCode Group and contributors
 *
 * MimicAPI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * MimicAPI is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with MimicAPI.  If not, see <http://www.gnu.org/licenses/>.
 */

package ru.endlesscode.mimic.api.system

/**
 * System that provides methods to work with players class systems.
 *
 * Before implementing run an eye over all default method implementations
 * and override all methods that works not properly for your case.
 *
 * @see [PlayerSystem] To read more about implementation.
 * @author Osip Fatkullin
 * @since 0.1
 */
interface ClassSystem : PlayerSystem {

    /**
     * Primary class of the player.
     *
     * It is called "primary" because some class systems can support many classes
     * for one player. Primary class can't be null if player has at least one class.
     *
     * @return Primary class name or `null` if player has not any classes.
     */
    @JvmDefault
    val primaryClass: String?
        get() {
            val classes = this.classes
            return if (classes.isEmpty()) null else classes[0]
        }

    /**
     * List of player classes.
     *
     * It is actual for systems which support many classes for one player.
     * If system doesn't support - it just returns [List] with one element.
     *
     * Never returns `null`, but can return empty [List].
     *
     * @return List of player system names
     * @throws IllegalStateException If player-related object not exists.
     */
    val classes: List<String>

    /** Returns `true` if player has any class. */
    @JvmDefault
    fun hasClass(): Boolean = primaryClass != null

    /** Returns `true` if player has any of [requiredClasses]. */
    @JvmDefault
    fun hasOneOfRequiredClasses(requiredClasses: List<String>): Boolean {
        return requiredClasses.any { this.hasRequiredClass(it) }
    }

    /** Returns `true` if player has all [requiredClasses]. */
    @JvmDefault
    fun hasAllRequiredClasses(requiredClasses: List<String>): Boolean {
        return requiredClasses.all { this.hasRequiredClass(it) }
    }

    /** Returns `true` if player has [requiredClass]. */
    @JvmDefault
    fun hasRequiredClass(requiredClass: String): Boolean = requiredClass in this.classes

    /** Factory of class systems. */
    class Factory(tag: String, constructor: (Any) -> ClassSystem)
        : SystemFactory<ClassSystem>(tag, constructor)
}
